roles/tests-integration/tests/common/sniffer.rs

```diff
@@ -0,0 +1,381 @@
+/// downstream and upstream roles in two separate FIFO queues. The messages are then forwarded to
+/// the target role without any modification.
+///
+/// It is useful for testing purposes, as it allows to assert that the roles have sent specific
+/// messages in a specific order and to inspect the messages details.
+#[derive(Debug, Clone)]
+pub struct Sniffer {
+    downstream: SocketAddr,
+    upstream: SocketAddr,
+    downstream_messages: MessagesAggregator,
+    upstream_messages: MessagesAggregator,
+}
+
+impl Sniffer {
+    /// Creates a new sniffer that listens on the given downstream address and connects to the
+    /// given upstream address.
+    pub async fn new(downstream: SocketAddr, upstream: SocketAddr) -> Self {
+        Self {
+            downstream,
+            upstream,
+            downstream_messages: MessagesAggregator::new(),
+            upstream_messages: MessagesAggregator::new(),
+        }
+    }
+
+    /// Starts the sniffer.
+    ///
+    /// The sniffer should be started after the upstream role have been initialized and is ready to
+    /// accept messages and before the downstream role starts sending messages.
+    pub async fn start(self) {
+        let (downstream_receiver, downstream_sender) =
+            Self::create_downstream(Self::wait_for_client(self.downstream).await)
+                .await
+                .expect("Failed to create downstream");
+        let (upstream_receiver, upstream_sender) = Self::create_upstream(
+            TcpStream::connect(self.upstream)
+                .await
+                .expect("Failed to connect to upstream"),
+        )
+        .await
+        .expect("Failed to create upstream");
+        let downstream_messages = self.downstream_messages.clone();
+        let upstream_messages = self.upstream_messages.clone();
+        let _ = select! {
+            r = Self::recv_from_down_send_to_up(downstream_receiver, upstream_sender, downstream_messages) => r,
+            r = Self::recv_from_up_send_to_down(upstream_receiver, downstream_sender, upstream_messages) => r,
+        };
+    }
+
+    /// Returns the oldest message sent by downstream.
+    ///
+    /// The queue is FIFO and once a message is returned it is removed from the queue.
+    ///
+    /// This can be used in order to assert the downstream role have sent a specific
+    /// message/message type.
+    pub fn next_downstream_message(&self) -> Option<(MsgType, AnyMessage<'static>)> {
+        self.downstream_messages.next_message()
+    }
+
+    /// Returns the oldest message sent by upstream.
+    ///
+    /// The queue is FIFO and once a message is returned it is removed from the queue.
+    ///
+    /// This can be used in order to assert the upstream role have sent a specific
+    /// message/message type.
+    pub fn next_upstream_message(&self) -> Option<(MsgType, AnyMessage<'static>)> {
+        self.upstream_messages.next_message()
+    }
+
```

Comment:
```suggestion

/// Matches against the oldest message sent by upstream.
///
/// The queue is FIFO and once a message is matched it is removed from the queue.
///
/// This can be used in order to assert the upstream sent a specific message type,
/// while ignoring specific message flags and fields.
```